### ISSUE TITLE
Add `title` field to embedded content

### ISSUE THREAD

**Issue by Norric1Admin** [ONSdigital/sdg-indicators#3083]
The embedded frames does not contain a title attribute to describe the iframes contents. Ensure that iframes contain non-empty title attributes that describe the contents of the frame (`title=”___”`).

```
<script type="text/javascript" src="https://pym.nprapps.org/pym.v1.min.js"></script> <script>document.querySelector("li.nav-item.embedded-map").addEventListener("click",function(){ var pymParent = new pym.Parent('embeddedmapview', 'https://www.ons.gov.uk/visualisations/dvc529/multiline/index.html', {});})</script> 
```
From: https://github.com/ONSdigital/sdg-data/blob/3-4-2-embed/meta/3-4-2.md

**Comment by Norric1Admin** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Accessibility tickets](https://trello.com/c/E40VA79G/549-accessibility-tickets)

**Comment by LucyGwilliamAdmin** [comment]
## Description

## Screenshots
**Before**

**After**


## Successfully tested on...
- [ ] Windows: IE11
- [ ] Windows: Chrome
- [ ] Windows: Firefox
- [ ] Windows: Edge
- [ ] macOS: Chrome
- [ ] macOS: Firefox
- [ ] iOS: Safari 9.3
- [ ] iOS: Chrome
- [ ] Samsung: Chrome
- [ ] Samsung: Samsung Internet

## User feedback and problems encountered...

**Comment by Norric1Admin** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Accessibility: Sprint 14](https://trello.com/c/80w0TtAV/586-accessibility-sprint-14)

**Comment by Norric1Admin** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Accessibility: Sprint 14](https://trello.com/c/80w0TtAV/586-accessibility-sprint-14)

**Comment by LucyGwilliamAdmin** [comment]
I think I've fixed this, needs to be tested with screen readers thought. Here's the [feature branch](https://sdg.mango-solutions.com/embeddedcontenttitles-deploy/)

**Comment by Norric1Admin** [comment]
We think a heading could be a better option (perhaps defined in a metadata field).

**Comment by LucyGwilliamAdmin** [comment]
I was going to try and do what Chris said but unfortunately by skills don't stretch that far. All I've done is make [this branch](https://github.com/ONSdigital/sdg-indicators/tree/add-embed-title) and I think the file that needs to be changed is [indicatorView.js](https://github.com/ONSdigital/sdg-indicators/blob/add-embed-title/_includes/assets/js/indicatorView.js) or at least that is where the title is added to the 'Title' tab - I'm thinking the embed would have a similar layout to the Table tab:
![image](https://user-images.githubusercontent.com/43404358/58168852-25e98180-7c87-11e9-898e-19c9d7e90645.png)
to
Title
Embedded content.
The title however would need to be added via a metadata field maybe 'embed_title'. Not sure if this would involve changing another file though.

I think this may be one for @SavvasStephanides ? @Norric1Admin

**Comment by Norric1Admin** [comment]
@dougmet  We'd like to add a custom title above embedded content, defined in the metadata. We have a feature branch but need to edit indicatorView.js.

**Comment by Norric1Admin** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Add `title` field to embedded content](https://trello.com/c/Af9LfuUh/607-add-title-field-to-embedded-content)

**Comment by dougmet** [comment]
Is this something that will go up to open-sdg?

**Comment by Norric1Admin** [comment]
Yes, all of the accessibility updates will become part of open-sdg (probably before @1.0.0).

**Comment by dougmet** [comment]
@Norric1Admin so the title needs to be in the iframe?

```html
<iframe src="..." width="100%" scrolling="no" marginheight="0" frameborder="0" height="675px" title="THETITLE">
```

**Comment by LucyGwilliamAdmin** [comment]
@dougmet I tried adding titles to the iframes but the screen reader didn't seem to be reading them out, it's as if they weren't being treated as a separate entity. Maybe it would work on the higher level but if not, maybe something like I described in my comment on 22nd May

**Comment by Norric1Admin** [comment]
@dougmet There are two options as I see it: i) a `title` field in the iframe, which is read by screen readers, or ii) an actual title/heading above the embedded content similar to the text above charts/tables (below). The choice really comes down to which you think would be easiest and quickest to implement. We have narrowed it down to [indicatorView.js](https://github.com/ONSdigital/sdg-indicators/blob/add-embed-title/_includes/assets/js/indicatorView.js), but are stuck on how to make this read from `graph_title`.

![image](https://user-images.githubusercontent.com/37659239/60030145-5fecdf80-969a-11e9-90a1-be56a1a63581.png)

**Comment by dougmet** [comment]
Scanning over the [pym.js](http://blog.apps.npr.org/pym.js/api/pym.js/1.3.2/module-pym.Parent.html) docs To just get a title attribute to the iframe you can do:

```html
<script type="text/javascript" src="https://pym.nprapps.org/pym.v1.min.js"></script> <script>document.querySelector("li.nav-item.embedded-map").addEventListener("click",function(){ var pymParent = new pym.Parent('embeddedmapview', 'https://www.ons.gov.uk/visualisations/dvc529/multiline/index.html', {"title": "THIS IS THE TITLE"});})</script> 
```

I'm not a huge fan of injecting JS like this. Also, for this method you couldn't use another metadata field.

Advantage: No code change requires because that script is stored in the metadata.

**Comment by dougmet** [comment]
It makes me wonder if the system for embedding should be (in the metadata)

```yaml
embedded_content_url: https://www.ons.gov.uk/visualisations/dvc529/multiline/index.html
embedded_content_title: A nice title
```

OR

```yaml
embedded_content_url: https://www.ons.gov.uk/visualisations/dvc529/multiline/index.html
embedded_content_config:
  title: A nice title
  name: a name attribute
```

**Comment by dougmet** [comment]
We can't use the caption as used for the table because the `<caption>` tag is specifically for tables

**Comment by brockfanning** [comment]
Hi all, as far as I know, the embed feature is not used by many (if any) other countries. So I think that if you are OK with the iframe-only approach, that could be added to open-sdg. I certainly support anything to spares countries from having to put javascript into their metadata.

Speaking of metadata, I could be wrong but I think that nested config are tricky (if not impossible?) in Prose.io. So maybe to keep Prose.io happy it's better to use Doug's first option above.

**Comment by dougmet** [comment]
Yes you're absolutely right about nested config.

**Comment by Norric1Admin** [comment]
Thanks for that @brockfanning. Think we should use the first option @dougashton-ons suggested, e.g. ...
```
embedded_content_url: https://www.ons.gov.uk/visualisations/dvc529/multiline/index.html
embedded_content_title: A nice title
```

**Comment by LucyGwilliamAdmin** [comment]
@dougmet is there anything I can do with this?

**Comment by Norric1Admin** [comment]
@LucyGwilliamAdmin I don't see an embed tab on the example above ([3.4.2](https://sdg.mango-solutions.com/embed-title/3-4-2/))?

**Comment by LucyGwilliamAdmin** [comment]
@Norric1Admin it's because I copied the latest version of indicator.html from open-sdg where the second Map tab hasn't been renamed to Embed. I will do that when it's all working properly

**Comment by Norric1Admin** [comment]
@LucyGwilliamAdmin I'm glad you were able to get this feature working. Will test it tomorrow.

**Comment by LucyGwilliamAdmin** [comment]
@SavvasStephanides I just tested this on iOS and the title was read out

**Comment by Norric1Admin** [comment]
@LucyGwilliamAdmin's commits above solve the problem with missing chart categories.

**Comment by Norric1Admin** [comment]
Adding a `tabindex` to make this title tabbable.
